### PR TITLE
Https via httpd and reverse proxy

### PR DESCRIPTION
**Pull request by razum2um**

### CODE REVIEW

File: smeagol.edn [hunk 1]
```diff
@@ -9,11 +9,12 @@
...
+   :orga-path "DomainDrivenArchitecture"
+   :repo-name "dda"
    :protocol :ssh
    :server-type :github}
  :server-fqdn "a.server.name"                                 ; the httpd server name
  :smeagol-passwd
     {:admin {:admin true,                                     ; smeagol users
              :email "admin@localhost",
-             :password {:plain "admin"}}}}
+             :password "admin" #_{:plain "admin"}}}}
```
Comment (razum2um) [inline smeagol.edn]: @jerger otherwise
```
ExceptionInfo Value does not match schema: {:smeagol-passwd {:admin {:password (not (instance? java.lang.String {:plain "admin"}))}}}  schema.core/validator/fn--7703 (core.clj:155)
```

Comment (jerger) [reply smeagol.edn]: We can 

1. use keypins resolving mechanism like `:password {:plain :envvar "YOUR_ADMIN_PASSWORD"}` - not sure whether this works,
2. we can add a security resolving like `{:local-edn-file {path "~/your-credential-file.edn" :key "your-key-in-map"}}` or 
3. you can use your personal config for your secure passwords and leave above config as instant running example in the repo.

Solution 3. will not require extension and work without additional infrastructure requirements - so I will vote for solution 3.

File: main/src/dda/pallet/dda_smeagol_crate/domain.clj [hunk 2]
```diff
@@ -20,11 +20,11 @@
     [dda.pallet.dda-smeagol-crate.domain.smeagol :as smeagol]
     [dda.pallet.dda-smeagol-crate.domain.git :as git]
     [dda.pallet.dda-smeagol-crate.domain.user :as user]
+    [dda.pallet.dda-smeagol-crate.domain.httpd :as httpd]
     [dda.pallet.dda-smeagol-crate.infra :as infra]))
 
 (def SmeagolDomain
-  {(s/optional-key :tomcat-xmx-megabyte) s/Num ;; necessary on level of domain?
-   (s/optional-key :smeagol-passwd) smeagol/SmeagolPasswd ;; reference from top level package is forbidden! moved to top-level
+  {(s/optional-key :smeagol-passwd) smeagol/SmeagolPasswd ;; reference from top level package is forbidden! moved to top-level
```
Comment (jerger) [inline main/src/dda/pallet/dda_smeagol_crate/domain.clj]: already thought about the comment in code?

File: main/src/dda/pallet/dda_smeagol_crate/domain/httpd.clj [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+  {:single-proxy
+   {:domain-name server-fqdn
+    :proxy-target-port proxy-port}})
+
+(s/defn
+  infra-configuration
+  [server-fqdn :- s/Str
+   proxy-port :- s/Str]
+  (httpd-domain/single-proxy-configuration
+   (domain-configuration server-fqdn proxy-port)))
```
Comment (jerger) [inline main/src/dda/pallet/dda_smeagol_crate/domain/httpd.clj]: on master I've moved infra-configuration to the app-ns.
My idea was: Domains responsibility is to create the foreign domain config & leave the infra creation to the app-ns.
But that's not symmetric to domain.smeagol - should we start finding reasons for one direction or just doing it all in app?

File: main/src/dda/pallet/dda_smeagol_crate/domain/smeagol.clj [hunk 4]
```diff
@@ -17,43 +17,71 @@
...
+  {:passwd {:env "SMEAGOL_PASSWD" :value (path-join root "passwd")}
+   :config-edn {:env "SMEAGOL_CONFIG" :value (path-join root "config.edn")}
+   ;; TODO git-crate infra result?!
+   :content-dir {:env "SMEAGOL_CONTENT_DIR" :value (path-join root "repo" "smeagol" "dda")}
+   :fonts {:env "TIMBRE_DEFAULT_STACKTRACE_FONTS" :value "{}"}
+   :log-level {:env "TIMBRE_LEVEL" :value ":info"}
+   :site-title {:env "SMEAGOL_SITE_TITLE" :value "DomainDrivenArchitecture"}
+   :default-locale {:env "SMEAGOL_DEFAULT_LOCALE" :value "en-GB"}
+   ;; TODO unify with httpd
+   :port {:env "PORT" :value "8080"}})
```
Comment (jerger) [inline main/src/dda/pallet/dda_smeagol_crate/domain/smeagol.clj]: why do we have these envs?

Wouldn't it be enough to put most of these values in the config file and just have one env containing config-files location?

File: main/src/dda/pallet/dda_smeagol_crate/domain/smeagol.clj [hunk 5]
```diff
@@ -17,43 +17,71 @@
...
-   {:name "PORT" :value "80"}])
+(def smeagol-releases "https://api.github.com/repos/DomainDrivenArchitecture/smeagol/releases")
 
-;; Review jem 2018_12_03: why not move tomcat into own ns ? - or remove it ;-)
-(s/defn tomcat-domain-configuration
-  [tomcat-xmx-megabyte]
-  {:app-server
-   {:xmx-megabyte tomcat-xmx-megabyte}})            ; e.g. 6072 or 2560
+(def ReleaseAsset
+  (open-schema {:browser_download_url s/Str :name s/Str :content_type s/Str :label (s/maybe s/Str)}))
```
Comment (jerger) [inline main/src/dda/pallet/dda_smeagol_crate/domain/smeagol.clj]: nice idea :-)

File: main/src/dda/pallet/dda_smeagol_crate/infra/schema.clj [hunk 6]
```diff
@@ -26,12 +26,12 @@
...
-   :smeagol-dir s/Str
-   :smeagol-passwd SmeagolPasswd
-   :smeagol-owner s/Str
-   :repo-download-source s/Str
-   :resource-locations {s/Keyword {:name s/Str :source s/Str :destination s/Str}}
-   :environment-variables [{:name s/Str :value s/Str}]})
+  {:passwd SmeagolPasswd
+   :owner s/Str
+   :uberjar SmeagolUberjar
+   :env {s/Keyword SmeagolEnv}})
```
Comment (jerger) [inline main/src/dda/pallet/dda_smeagol_crate/infra/schema.clj]: let's kill infra/schema and move schema definition directly into infra/smeagol ...
All smeagol definition's should only be used by smeagol ns - if not, that's clearly sth. we should think about ;-)

File: main/src/dda/pallet/dda_smeagol_crate/infra/smeagol.clj [hunk 7]
```diff
@@ -17,78 +17,103 @@
...
+(s/defn download-uberjar
+  [owner :- s/Str
+   uberjar :- schema/SmeagolUberjar]
+  (let [{:keys [path url size]} uberjar
+        all-facts (crate/get-settings
+                   fact/fact-facility
+                   {:instance-id (crate/target-node)})
+        file-fact (file-fact-keyword all-facts)
+        fact-path (path-to-keyword path)]
+    (actions/plan-when (let [{:keys [fact-exist? fact-size-in-bytes] :as actual}
```
Comment (jerger) [inline main/src/dda/pallet/dda_smeagol_crate/infra/smeagol.clj]: Let's not consider state here ... to keep things simple ... as we tend to recreated servers over modify existing ones ...

You can use "install-phase" for jar download.

The "config-phase" is intended to do small config changes in a continuous way ... (if it's too expensive to recreate a server and configs can be easily be overwritten without state & data loss concerns ...)

File: main/src/dda/pallet/dda_smeagol_crate/domain/smeagol.clj (hunk 4 above)
Comment (razum2um) [reply main/src/dda/pallet/dda_smeagol_crate/domain/smeagol.clj]: I got a bit stuck with the smeagol setup, it was complaining about missing `:site-title` and `:default-locale` (wtf for me). Same after this settings, need to investigate.

First of all I'd rework how smeagol accepts configuration first: rewrite using keypin or smth like that to make it more transparent, without tramsform-map etc, ok?

Then, I'd prefer everything as simple values in ENV rather than a config since we (other guys) might reuse samples to start docker containers which are generally not ok with passing config path via ENV, but rather passing just values (esp. given they're plain strings, not structures) via `-e`. 

TLDR: I think sometimes in the future we'll start docker containers as well, so I'd insist on ENV-first approach

File: main/src/dda/pallet/dda_smeagol_crate/infra/schema.clj (hunk 6 above)
Comment (razum2um) [reply main/src/dda/pallet/dda_smeagol_crate/infra/schema.clj]: ok, will do

File: main/src/dda/pallet/dda_smeagol_crate/infra/smeagol.clj (hunk 7 above)
Comment (razum2um) [reply main/src/dda/pallet/dda_smeagol_crate/infra/smeagol.clj]: I'd not agree since I don't think it's much complicated, besides, I very like the idea, that the second run of any phase should be as instant as possible, Ideally we should gather even more info via serverspec to get as optimised plan as possible and insist to leverage fact data advertising serverspec facilities by the way.

> we tend to recreated servers over modify existing ones
right now I'm talking about general approach, that's why no "we" - I think people in common (before containers/aws-AMI era) tend to modify servers, wdyt?

Still, the very explicit manner of how this is written bothers me as well, and I already have a doc where I collect things I don't like about pallet/dda - it's one of them. I think this can and should be encapsulated into a simpler/clearer form

I think we'd better to prevent NPEs like you solved and make usage less error-prone. e.g. will we discuss some ideas how to compose `with-..` fns etc?

File: smeagol.edn (hunk 1 above)
Comment (razum2um) [reply smeagol.edn]: imo 1 is best if it works out of the box (haven't tried yet)
3 is ok, but it requires careful watching for domain changes etc, besides, it's a origin of a "works for you, doesn't work for me - what's your config - ah, need to update config"

File: main/src/dda/pallet/dda_smeagol_crate/domain/httpd.clj (hunk 3 above)
Comment (jerger) [reply main/src/dda/pallet/dda_smeagol_crate/domain/httpd.clj]: sorry missed the context here ... your solution is fine :-)

File: main/src/dda/pallet/dda_smeagol_crate/domain/smeagol.clj (hunk 4 above)
Comment (jerger) [reply main/src/dda/pallet/dda_smeagol_crate/domain/smeagol.clj]: 12factor apps?
Some envs make no sense to expose them - even if we will use docker images as artefacts (think about the git repo location). Lets first bring a simple solution to live and add the missing parts as soon as the usecase occurs.

File: main/src/dda/pallet/dda_smeagol_crate/infra/smeagol.clj (hunk 7 above)
Comment (jerger) [reply main/src/dda/pallet/dda_smeagol_crate/infra/smeagol.clj]: Every fact add state to our code & makes testing more difficult & expensive. I propose to use as view state as possible ...

There are many examples that too many state is not working in the generation world:
* think about the mda roundtrip discussion ...
* puppet's CM and it's dificulties on state / historyless one-shoot setup / aso

Let's focus on cloud & immutable images ... makes live easier a lot.

File: main/src/dda/pallet/dda_smeagol_crate/infra/smeagol.clj [hunk 8]
```diff
@@ -63,29 +80,34 @@
...
-                       (actions/remote-file path
-                                            :url url
-                                            :owner owner))))
+    ;                     (not (and fact-exist? (= fact-size-in-bytes size)))
+       (actions/packages :aptitude ["curl"])
+       (actions/directory "/usr/local/lib/smeagol"
+                          :owner owner
+                          :group "users"
+                          :mode "755")
+       (actions/remote-file "/usr/local/lib/smeagol/smeagol-standalone.jar"
```
[the remark this answers is not in the document]
Comment (jerger) [reply main/src/dda/pallet/dda_smeagol_crate/infra/smeagol.clj]: Versioning in the immutable image world is same as in versioned jars.
U can use which image to use instead of making one image able to use multiple versions.

In the java universe I know only eclipse supporting a versioned plugin mechanism ... and I know the cost in terms of eclipse plugin complexity very well ...

File: main/src/dda/pallet/dda_smeagol_crate/infra/smeagol.clj [hunk 9]
```diff
@@ -109,11 +131,16 @@
...
+  [config :- SmeagolInfra]
   (let [{:keys [uberjar passwd owner env]} config]
     ;; TODO shared :owner like `with-action-options`?!
-    (create-smeagol-config config)
-    (create-passwd (:passwd env) owner passwd)
     (download-uberjar owner uberjar)
     (initd-script env uberjar)
     (smeagol-service (vals env) uberjar)))
+
+(s/defn configure-smeagol
```
[the remark this answers is not in the document]
Comment (jerger) [reply main/src/dda/pallet/dda_smeagol_crate/infra/smeagol.clj]: As I advocate for immutability arguing against is hard :-)

There are some use cases we will need mutability like:
* add or remove users to whole infrastructure (on-offboarding process) (relevant for user crate)
* modify more mutable setups like desktop-distributions (relevant for git & vm & ide crates)

If you execute the uberjar we will execute phasese "settings-install-config" but you will be able to run "settings-config" only if you choose the "--conigure" option.

File: main/src/dda/pallet/dda_smeagol_crate/infra/smeagol.clj (hunk 8 above)
[the remark this answers is not in the document]
Comment (razum2um) [reply main/src/dda/pallet/dda_smeagol_crate/infra/smeagol.clj]: So you see the primary and almost single purpose of crates in building images (AMI/whatever)? Is it not intended to be run against a single running (mutable) server?

File: main/src/dda/pallet/dda_smeagol_crate/infra/smeagol.clj [hunk 10]
```diff
@@ -63,29 +80,34 @@
...
+    ;(actions/plan-when (let [{:keys [fact-exist? fact-size-in-bytes] :as actual}
+    ;                         (fact-path (:out @file-fact))
                          ;; (logging/info (pr-str {:actual actual :expected uberjar}))
-                         (not (and fact-exist? (= fact-size-in-bytes size))))
-                       (actions/remote-file path
-                                            :url url
-                                            :owner owner))))
+    ;                     (not (and fact-exist? (= fact-size-in-bytes size)))
+       (actions/packages :aptitude ["curl"])
+       (actions/directory "/usr/local/lib/smeagol"
```
[the remark this answers is not in the document]
Comment (razum2um) [reply main/src/dda/pallet/dda_smeagol_crate/infra/smeagol.clj]: ok, I tend to agree here just for the sake of simplicity... although 2 thoughts in depth:

> separates executable (/usr/lib or /opt) & data 

that's good & well established, but I think grouping apps/tasks by purpose and *isolation* tasks from each other is also important. that was the basic point of containers but in their's absence I'd separate users to isolate both task's code, data, any fs locations from each other completely (think jails missing in linux etc)

> home/user paths are not guaranteed to be available

good point. but I worry about it: basically `:owner owner` file mode and more importantly starting a service under `owner` is a *hidden* coupling between this crate and user existence anyway. I think there's no way to escape it unless we'll use root. Otherwise we can make this dependency *explicit*

`with-user` in `app` almost does express the dep *but* as i realise now there's no order inside a phase (am I correct?) and we *hide* dependency using different phases. should we always have an extra phase to declare the strict order? that's why I feel strange about multiple phases :\

File: main/src/dda/pallet/dda_smeagol_crate/infra/smeagol.clj [hunk 11]
```diff
@@ -63,29 +80,34 @@
...
-                             (fact-path (:out @file-fact))]
+    ;(actions/plan-when (let [{:keys [fact-exist? fact-size-in-bytes] :as actual}
+    ;                         (fact-path (:out @file-fact))
                          ;; (logging/info (pr-str {:actual actual :expected uberjar}))
-                         (not (and fact-exist? (= fact-size-in-bytes size))))
-                       (actions/remote-file path
-                                            :url url
-                                            :owner owner))))
+    ;                     (not (and fact-exist? (= fact-size-in-bytes size)))
+       (actions/packages :aptitude ["curl"])
```
[the remark this answers is not in the document]
Comment (razum2um) [reply main/src/dda/pallet/dda_smeagol_crate/infra/smeagol.clj]: made a todo for me for the future

File: main/src/dda/pallet/dda_smeagol_crate/infra/smeagol.clj (hunk 9 above)
[the remark this answers is not in the document]
Comment (razum2um) [reply main/src/dda/pallet/dda_smeagol_crate/infra/smeagol.clj]: Hm, I noticed that maybe we could apply these tools to stateful/mutable servers, but here I'm talking about several phase and don't fully understand how it's connected to immutability :)

> only if you choose the "--conigure" option

yeah, this makes me crazy... install will run against vm which will be the image then. but configure is run against running mutable servers, correct?

File: main/src/dda/pallet/dda_smeagol_crate/infra/smeagol.clj (hunk 8 above)
[the remark this answers is not in the document]
Comment (jerger) [reply main/src/dda/pallet/dda_smeagol_crate/infra/smeagol.clj]: we tend to setup new server for new rollout s ... so intention is mainly build images but will also work on (mutable) servers as well - if u throw away the old ones & migrates to newly set up ..

File: main/src/dda/pallet/dda_smeagol_crate/infra/smeagol.clj (hunk 9 above)
[the remark this answers is not in the document]
Comment (jerger) [reply main/src/dda/pallet/dda_smeagol_crate/infra/smeagol.clj]: Or think about the aws-rds service. Image may represent a postgres-db-server for example, but it may be cool if your customer can bring his configuration to instantiated server by filling some html forms and apply configuration in a later point in time ...

File: main/src/dda/pallet/dda_smeagol_crate/infra/smeagol.clj (hunk 10 above)
[the remark this answers is not in the document]
Comment (razum2um) [reply main/src/dda/pallet/dda_smeagol_crate/infra/smeagol.clj]: so maybe I'll add todo for making clone in install phase of git-crate and configure will do same but also a pull, ok?

File: main/src/dda/pallet/dda_smeagol_crate/domain.clj [hunk 12]
```diff
@@ -51,8 +51,8 @@
 (s/defn ^:always-validate
   httpd-domain-configuration
   [domain-config :- SmeagolDomainResolved]
-  (let [{:keys [server-fqdn proxy-port settings] :or {proxy-port "8080"}} domain-config]
-    (httpd/domain-configuration server-fqdn proxy-port (contains? domain-config :settings) settings)))
```
Comment (razum2um) [inline main/src/dda/pallet/dda_smeagol_crate/domain.clj]: @jerger btw, this is exaclty what I warned about: you have an own config with `:settings` and I have trouble to run out-of-the-box because `settings` was passed as `nil` and it violates `VhostSettings`

File: main/src/dda/pallet/dda_smeagol_crate/domain/smeagol.clj [hunk 13]
```diff
@@ -81,7 +84,7 @@
 (s/defn smeagol-infra-configuration
   [facility :- s/Keyword
    repo-name :- s/Str
-   passwd :- SmeagolPasswd]
+   passwd :- SmeagolPasswdResolved]
```
Comment (razum2um) [inline main/src/dda/pallet/dda_smeagol_crate/domain/smeagol.clj]: @jerger smth is definetly strange happening - you don't notice some schema violations like this (at this point passwords are already inlined to strings and I already complaned about it in first comments.

same for https://github.com/DomainDrivenArchitecture/dda-serverspec-crate/pull/17 <- it's definitely broken output schema, why you don't merge it?

File: main/src/dda/pallet/dda_smeagol_crate/domain.clj (hunk 12 above)
Comment (jerger) [reply main/src/dda/pallet/dda_smeagol_crate/domain.clj]: OpenSchema ? Yes - but this happens only if we validate on sub-ns ... our convention is to validate only on app / domain / infra level. Other spec are validated in tests only ...

On the midterm run I want to switch to spec ... and we will follow the open schema idea on this path ...

File: main/src/dda/pallet/dda_smeagol_crate/domain/smeagol.clj (hunk 13 above)
Comment (jerger) [reply main/src/dda/pallet/dda_smeagol_crate/domain/smeagol.clj]: So we should improve testing here so we will recognize not matching schema ...

Comment (jerger) [reply main/src/dda/pallet/dda_smeagol_crate/domain/smeagol.clj]: Reg. dda-serverspec: At the moment I focus on dda-smeagol-crate. As soon as we finished I will review other PRs.

I would propose to add tests showing the issue u found ... so we can improve quality enduring ...

File: main/src/dda/pallet/dda_smeagol_crate/infra/smeagol.clj (hunk 11 above)
[the remark this answers is not in the document]
Comment (razum2um) [reply main/src/dda/pallet/dda_smeagol_crate/infra/smeagol.clj]: @jerger was debugging pallet and found that it executes
```
if hash curl 2>&-; then
curl -o ...
else
if hash wget 2>&-; then
wget -O ...
else
echo No download utility available
exit 1
```
:\

File: main/src/dda/pallet/dda_smeagol_crate/infra/smeagol.clj (hunk 8 above)
[the remark this answers is not in the document]
Comment (razum2um) [reply main/src/dda/pallet/dda_smeagol_crate/infra/smeagol.clj]: @jerger about backups/symlinks - Hugo already did it: what `remote-file` does a copy of everything under `/var/lib/pallet` md5-ing everything and backup-ing 6 previous versions. Did you know? :)

File: main/src/dda/pallet/dda_smeagol_crate/domain/smeagol.clj (hunk 13 above)
Comment (razum2um) [reply main/src/dda/pallet/dda_smeagol_crate/domain/smeagol.clj]: > add tests showing the issue u found
Saw you already turned on schema validation there, thanks
